### PR TITLE
And we need the new catalog layout as well

### PR DESCRIPTION
**Pull request by quintesse**
See #148

### CODE REVIEW

**Comment by gastaldi** [comment]
Is this synchronized with the latest changes in master from today?

**Comment by quintesse** [comment]
I first need to finish the `environments` stuff, so I can merge the data from `next` and from whatever is in production

**Comment by gastaldi** [comment]
Can't we automatically assume that if a cluster type is not listed under `runsOn` it does not run?

**Comment by quintesse** [comment]
It supports both. But think about it: what names do we choose for those clusters? What name do we give a local minishift? What name do we give "int-eng", etc? There's more than just "free" and "pro". So the moment you say "runsOn: free, pro" what do we do with the others? So in my opinion it's easier to say where it does _not_ run, eg: "doesNotRunOn: free". Also, how do you say "it doesn't run anywhere" with "runsOn"?

**Comment by heiko-braun** [comment]
Please don't change this at good and think of backwards compatibility and deprecation as well. The booster releases are costly for the runtime teams, especially the product booster ones.

**Comment by quintesse** [comment]
@heiko-braun I've been making all the changes necessary myself to make sure there won't be any work for you guys. Unless you have a bunch of private catalogs that would need to be updated as well? (In which case perhaps a simple script could be written that does that work for you)

I think backward compatibility is good of course, but it shouldn't prevent us from making necessary changes to take this product from something that was written to get a very specific limited set of functionality out before summit and take it to something that can be an actual community project.

**Comment by gastaldi** [comment]
We'll probably want to support boosters from another public git repositories (including those from bitbucket, gitlab, etc).

I think it's a good idea to rename `githubRepo` to `gitRepo` and add the github.com prefix to the existing repositories

**Comment by quintesse** [comment]
Sounds like a good idea

**Comment by gastaldi** [comment]
Specially if we are dealing with private repositories too

**Comment by gastaldi** [comment]
Boosters used to have an ID, based on YAML file name in this catalog. Is that no longer necessary?

**Comment by quintesse** [comment]
@gastaldi from the catalog service commit message:

 >  - Booster ids are now generated by using their full path relative to the catalog root folder,
   meaning that it's not necessary anymore to create unique names for each booster decriptor
   .yaml file in the catalog.

😄

**Comment by quintesse** [comment]
> I think it's a good idea to rename githubRepo to gitRepo and add the github.com prefix to the existing repositories

Ok, I found a slight issue with this. The tests assume they can just fire up a local Git server and use those URLs (because right now they are relative). If we put full prefix in there we need something a bit more involved to make the tests work. Any ideas?

Btw: would you put in the full URL including `https://` or just `github.com\`?

**Comment by gastaldi** [comment]
I'd put the full URL including `https://`. The idea was to just use that for the git clone command (which could be a SSH URL too)

**Comment by quintesse** [comment]
True, but might that not make things like figuring out the correct authentication to use more difficult?

**Comment by gastaldi** [comment]
We could bundle a `.gitconfig` with our backend service perhaps. How would that be less difficult with the current approach?

**Comment by quintesse** [comment]
Because right now we determine the connection mechanism (which is https), if someone by accident puts in a ssh url I'm pretty sure the current authentication would fall (as it would go to the local filesystem looking for a key)

**Comment by quintesse** [comment]
But perhaps it's not a big problem and it's just the catalog author's responsibility to make sure all urls are correct

**Comment by gastaldi** [comment]
Agreed. I'd say let's use the full URL and handle each corner case separately.

**Comment by Ladicek** [comment]
FYI, the Swarm product version `7.0.0.redhat-6` is wrong, not sure where that comes from, the correct product version is `7.0.0.redhat-8`.

Couple more random thoughts:
- Not sure about the "metadata" key, it's all metadata, right? :-)
- After this, the `.openshiftio/booster.yaml` files from the boosters are ignored and can be deleted? YAY!
- The `runsOn` and `doesNotRunOn` keys -- what's the priority when both are present? Wouldn't a boolean expression language (just cluster names, `&&`, `||`, `!` and parens) be easier to understand? Or a regexp?

**Comment by gastaldi** [comment]
>Not sure about the "metadata" key, it's all metadata, right? :-)

Agreed, I think we could remove the metadata key

>The runsOn and doesNotRunOn keys -- what's the priority when both are present? Wouldn't a boolean expression language (just cluster names, &&, ||, ! and parens) be easier to understand? Or a regexp?

An expression or a regexp would be awesome, considering that our options won't be too abroad. Maybe even introducing functions or variables to pointing to values in the same YAML file

**Comment by quintesse** [comment]
@Ladicek could be from an older version of the catalog, I need to update them once the code has been merged.

 - `metadata`: this is basically copied from the OpenShift templates. The toplevel properties are used by the catalog service while "metadata" is just whatever you can think of, the catalog service will only store them but ignore them completely.
 - Yes! :)
 - True, but I wanted something simple, an expression language isn't simple. Perhaps the two could be put into one with `!` signifying "not". But then "none" becomes `!*` which isn't very nice. But it's definitely an option.

@gastaldi I don't see why we should get rid of "metadata"? Mixing them in with the top-level properties removes the indication that they are user-defined properties that have no significance for the catalog service itself.

**Comment by gastaldi** [comment]
@quintesse I think @Ladicek's point is exactly this: to have top level properties and metadata as a single entity. 

Although I think it's a good idea, maybe there are cases I haven't covered. I think we assume the top level properties as required right?

**Comment by quintesse** [comment]
> An expression or a regexp would be awesome, considering that our options won't be too abroad. Maybe even introducing functions or variables to pointing to values in the same YAML file

Right now basically there are only _two_ things we need this for: "openshift-online-free" and "none whatsoever" (= zip only download). Going for expressions with variables is overkill at this point in time. If we ever need that we could easily add it.

KISS: Don't create complex stuff before you need it :)

**Comment by gastaldi** [comment]
>Mixing them in with the top-level properties removes the indication that they are user-defined properties that have no significance for the catalog service itself.

Not sure I agree with that. The information will be used somewhere (when filtering), otherwise it doesn't make sense to exist ;)

**Comment by gastaldi** [comment]
>Going for expressions with variables is overkill at this point in time. If we ever need that we could easily add it.
>KISS: Don't create complex stuff before you need it :)

Agreed, that was just an idea, not necessary a need ;)

**Comment by quintesse** [comment]
>  I think @Ladicek's point is exactly this: to have top level properties and metadata as a single entity.

Yes, and I'm saying that it removes the distinction I think is important: that the toplevel properties are the ones used by the service it self (some of them required , some of them not) while the rest are completely undefined, you can put the works of Shakespeare there if you want. Separating the two seems good design to me. Being facetious and saying "everything is meta" is like saying we can just use global variables in code ;)

**Comment by gastaldi** [comment]
>Perhaps the two could be put into one with ! signifying "not". But then "none" becomes !* which isn't very nice. But it's definitely an option.

I like that option. It keeps everything under a single attribute (instead of `runsOn` and `doesNotRunOn`).

**Comment by quintesse** [comment]
> Agreed, that was just an idea, not necessary a need ;)

I do agree that I don't like the `doesNotRunOn` much though. So perhaps a single `runsOn` with `!` for "not"? What do you guys think?

We'd theoretically get things like:

```
metadata:
   runsOn:
   - production
   - !free
```

And of course the somewhat ugly "nowhere":

```
metadata:
   runsOn:
   - !*
```

**Comment by quintesse** [comment]
> I like that option. It keeps everything under a single attribute (instead of runsOn and doesNotRunOn).

Ok, will do that then

**Comment by gastaldi** [comment]
>you can put the works of Shakespeare there if you want

Yes, but if that's never used, why would anyone do that?

**Comment by Ladicek** [comment]
Wow :-)

Re `runsOn` and `doesNotRunOn`: agree that adding an expression language right now would be an overkill. At the same time, splitting the rules into two keys (`runsOn` and `doesNotRunOn`) is harder to understand (I don't think my question about priority was answered really :-) ) and prevents doing something more complex in the future. So I'd agree with putting it all into `runsOn` with some simple form of negation.

Re `metadata`: sorry, I fail to see the difference. Everything that's currently in the YAML is used by the Launcher in some way or another, no?

**Comment by quintesse** [comment]
> Yes, but if that's never used, why would anyone do that?

But we _do_ use it. The user uses the metadata section to put the information _they_ care about. In the RHOAR Launcher we ourselves for example care about `runsOn`, but that's completely orthogonal to the booster service, it doesn't need to know anything about that attribute, and doesn't know what to do with it.

> Re metadata: sorry, I fail to see the difference. Everything that's currently in the YAML is used by the Launcher in some way or another, no?

@Ladicek The Launcher is the product _we_ build on top of components like the Booster Catalog Service. And at some point this is all supposed to become some community project. A community project that's filled everywhere with RHOAR-specific things (Missions, Runtimes, Versions, OpenShift cluster names) is never going to get much traction IMO. So I've been moving to a split of base code that doesn't have anything related to the ROAR product. And most of the code was already like that tbh, most of the RHOAR specific code lives in the "addon", I just moved _everything_ related to RHOAR to the addon.

So no, the Booster Catalog Service only cares about the toplevel elements of the .yaml file, it doesn't care at all about the metadata section. The Launcher then takes that information and does extra things with it, eg: filtering (which Boosters to show), build options (how to build the project), etc

Now, if we wouldn't be trying to make this usable as a community project I'd agree and say: let's just make a big bag of properties, who cares, as long as it works.

**Comment by Ladicek** [comment]
@quintesse I see. Thanks.

**Comment by metacosm** [comment]
How about a "special" `none` value instead of `!*`?

**Comment by quintesse** [comment]
@metacosm it's definitely a possibility. I guess nobody is going to create a server type named "none" ;)

File: rest-http-secured/nodejs/booster.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+source:
+  git:
+    url: https://github.com/bucharest-gold/nodejs-rest-http-secured
+    ref: master
+name: Node.js - REST and RH SSO
+description: A simple HTTP application using Node.js secured by RH SSO
+metadata:
+  runsOn: "none"
```
Comment (gastaldi) [inline rest-http-secured/nodejs/booster.yaml]: is the `"` necessary?

Comment (quintesse) [reply rest-http-secured/nodejs/booster.yaml]: Nope, the latest version doesn't have that anymore :)

**Comment by quintesse** [comment]
@gastaldi I created a quick & dirty converter for the catalog. It was too much manual work which I would then never be really 100% was correct. So the current contents has been generated with that tool. (https://github.com/quintesse/launcher-catalog-converter)

**Comment by gastaldi** [comment]
Not in Kotlin? Shame on you :D

**Comment by gastaldi** [comment]
JK, good job!

**Comment by quintesse** [comment]
> Not in Kotlin? Shame on you :D

Wdym? It _is_ in Kotlin ;)

**Pull request merged**